src/V3Timing.cpp

```diff
@@ -0,0 +1,785 @@
+
+        // Create a fork var if it's not join_none
+        AstVarScope* forkVscp = nullptr;
+        if (!nodep->joinType().joinNone()) {
+            auto* const forkVarp
+                = new AstVar{flp, VVarType::BLOCKTEMP, forkName + "__handle", m_forkDtp};
+            forkVarp->funcLocal(true);
+            forkVscp = new AstVarScope{forkVarp->fileline(), m_scopep, forkVarp};
+            m_scopep->addVarp(forkVscp);
+            nodep->addHereThisAsNext(forkVarp);
+        }
+
+        // Transform each statement into a CFunc
+        auto stmtp = nodep->stmtsp();
+        vluint32_t procCount = 0;
+        while (stmtp) {
+            AstCFunc* const cfuncp
+                = new AstCFunc{stmtp->fileline(), forkName + "__" + std::to_string(procCount++),
+                               m_scopep, "VlCoroutine"};
+            m_scopep->addActivep(cfuncp);
+            auto* const ccallp = new AstCCall{stmtp->fileline(), cfuncp};
+            stmtp->replaceWith(ccallp);
+
+            if (auto* const beginp = VN_CAST(stmtp, Begin)) {
+                stmtp = beginp->stmtsp()->unlinkFrBackWithNext();
+                beginp->deleteTree();
+            }
+            cfuncp->addStmtsp(stmtp);
+
+            // Add a fork done() statement if it's not join_none
+            if (forkVscp) {
+                auto* const donep = new AstCMethodHard{
+                    flp, new AstVarRef{flp, forkVscp, VAccess::READWRITE}, "done"};
+                donep->dtypeSetVoid();
+                donep->statement(true);
+                cfuncp->addStmtsp(donep);
+            }
+
+            // We need to pass all locals used by the statement to the CFunc
+            std::map<AstVarScope*, AstVarScope*>
+                locals;  // Map from var accessed by process to func-local var
+            // Gather local vars referred to in our statement
+            cfuncp->foreach<AstVarRef>([&](AstVarRef* refp) {
+                if (refp->varp()->isFuncLocal())
+                    locals.insert(std::make_pair(refp->varScopep(), nullptr));
+            });
+
+            // Create a list of args from locals and put in the CFunc
+            for (auto& p : locals) {
+                auto* const varscp = p.first;
+                auto* const varp = varscp->varp()->cloneTree(false);
+                varp->direction(VDirection::INPUT);
+                cfuncp->addArgsp(varp);
+                AstVarScope* const newvscp = new AstVarScope{varp->fileline(), m_scopep, varp};
+                m_scopep->addVarp(newvscp);
+                p.second = newvscp;
+                ccallp->addArgsp(new AstVarRef{stmtp->fileline(), varscp, VAccess::READ});
+            }
+
+            // Remap varrefs in the CFunc
+            cfuncp->foreach<AstVarRef>([&](AstVarRef* refp) {
+                if (auto* const newvscp = locals[refp->varScopep()]) {
+                    refp->varScopep(newvscp);
+                    refp->varp(newvscp->varp());
+                }
+            });
```

Comment:
Iteration order of `locals` is non-deterministic so your arg list will be unstable. Also you should be able to do this (the foreach/for/foreach triplet) in a single foreach traversal. You can use the bool in the return value of `emplace` to notice it's the first time you added the VarScope and do what you do in the `for` loop, and immediately remap the reference.